text/0021-state-pruning-and-fast-syncing.md

```diff
@@ -0,0 +1,240 @@
+
+By deciding to "cut" the block chain at a specific parent, we can apply a similar garbage collection algorithm:
+
+1. Decide which blocks to keep and add them to the "objects to keep" set (*i.e.* a starting block and all descendents).
+1. For each block, visit all of its dependencies (deploys, finality signatures) and add them to an "objects to keep" set.
+1. Visit each block, deploy and finality signature, remove them if not inside the "objects to keep" set.
+
+In a manner similar to the trie store, we can employ a mark-and-sweep version of this pruning algorithm as well.
+
+This again reduces the problem on selecting which blocks to keep.
+
+### Pruning summary
+
+Both trie-store as well as storage can be trimmed down by garbage collection, with savings in the latter probably being more substantial than the former.
+
+In both cases the problem can reduced to identifying which root hashes or blocks must be kept and which can be discarded. Once this information is available, garbage collection becomes straight forward.
+
+An "archival node" is a node on which garbage collection is either disabled or restricted to non-transitive state information, *i.e.* it keeps all root hashes except those created in the midst of execution.
+
+### Fast-syncing
+
+Fast syncing, while beneficial in its own right, is also of interest from a storage perspective, as being able to recover information quickly from the network or external sources influences our decisions about what information must be kept.
+
+The "fast" part of fast-syncing, as compared to the status quo, is realized from two factors, namely
+
+* **side-loading** data from outside the network through external sources and
+* **transfering** state, namely trie-store state, which is currently not transferrable at all.
+
+### Side-loading
+
+We create a common trait for Blocks, Deploys or any other transferable object that states how it can be serialized and defines a hash function for its serialized representation, as well as a key to separate the hash space. As an example, a block whose hash is `5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2` would be adressable as `blockv1_5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2`. The `blockv1` denotes a complete scheme, that is it also pins the serialization format. While it is not necessary to separate the hash value space to avoid collisions, the prefix allows for multiple representations, should the format of a serialization or an internal data structure change at some point. We refer to data structures that satisfy this condition as *objects* from here on out.
```

Comment:
I don't see a need for the prefix.  Are you trying to support retrospectively changing serialization formats for historic blocks?

Since we have a protocol version, I'd expect we'd be able to define all necessary systemic changes via that one versioning mechanism.  So, if a block was stored under protocol version 1.0.0, we definitively know the composition/structure of the block and its serialization format.

Reply:
I'm not seeing the need either.  

I would assume every data fetch would starts with the getting a block, and the block itself has protocol version field.  As long as we guarantee that the protocol version will always have a predictable/fixed location and format in the binary format a block, all subsequent fetches,  be they deploys or trie nodes, will be able to use the block's protocol version to determine which deserializer to use.

Reply:
> I don't see a need for the prefix. Are you trying to support retrospectively changing serialization formats for historic blocks?

Data at rest cannot be assumed to be updated, I am thinking of external stores as basically write-only (e.g. a huge, historical S3 bucket). It is conceivable that data might be added twice in different formats, should a need arise to change it.

If we want to version instead, there is an alternative scheme: Store every object at an endpoint like

```
GET /objects/v1.0.0/5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2
```

I am assuming that a change of serialization format automatically increments a major version, so we might get away with

```
GET /objects/v1/5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2
```

Now if the format changes, we would have to create a new object at 

```
GET /objects/v2/5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2
```

via a conversion script. For objects whose serialization format hasn't changed, this would hopefully just be another URL where it is exposed on, without changing the underlying data.

> I would assume every data fetch would starts with the getting a block, and the block itself has protocol version field. As long as we guarantee that the protocol version will always have a predictable/fixed location and format in the binary format a block, all subsequent fetches, be they deploys or trie nodes, will be able to use the block's protocol version to determine which deserializer to use.

I believe that's just inlining the otherwise external version tag. It relies on the fact that a version tag is found at the same place in every object, which I find a bit brittle - should we ever decide to change things in the future, this might cause problems.

There is also the case where a serialization format for an object changes, without its hash (based on values) changing, so these objects might be transferred across the wire differently even though they are the same.